content/blog/2025-04-02-valkey-bloom.md

```diff
@@ -0,0 +1,188 @@
+
+## Data type overview
+
+The “Bloom Object” is the main bloom data type structure. This is what gets created with any bloom filter creation command and this structure can act either as a “scaling bloom filter” or “non scaling bloom filter” depending on the user configuration. It consists of a vector of “Sub Filters” with length >= 1 in case of scaling and only 1 in case of non scaling.
+
+The “Sub Filter” is an inner structure which is created and used within the “Bloom Object”. It tracks the capacity and number of items added, and an instance of a Bloom Filter (of the specified properties) from an external crate named bloomfilter.
+
+<img src="/assets/media/pictures/bloomfilter_datatype.png" style="width: 75%;" alt="bloom filter data type">
+
+**Non Scaling**
+
+When non-scaling filters reach their capacity, if a user tries to add a new/unique item, an error is returned.
+You can create a non scaling bloom filter using `BF.RESERVE` or `BF.INSERT` commands.
+Example:
+```
+BF.RESERVE <filter-name> <error-rate> <capacity> NONSCALING
+``` 
+
+**Scaling**
+
+When scaling filters reach their capacity, if a user adds an item to the bloom filter, a new sub filter is created and added to the vector of sub filters. This new bloom sub filter will have a larger capacity (previous_bloomfilter_capacity * expansion_rate of the bloom filter).
+When checking whether an item exists on a scaled out bloom filter (`BF.EXISTS`/`BF.MEXISTS`), we look through each filter (from oldest to newest) in the sub filter vector and perform a check operation on each one. Similarly, to add a new item to the bloom filter, we check through all the filters to see if the item already exists and the item is added to the current filter if it does not exist.
+Any default creation as a result of `BF.ADD`, `BF.MADD`, `BF.INSERT` will be a scalable bloom filter.
+
+**Common Bloom filter properties**
+
+Capacity - The number of unique items that can be added before a bloom filter scales out occurs (in case of scalable bloom filters) or before any command which inserts a new item is rejected (in case of non scalable bloom filters).
+
+False Positive Rate (aka Error Rate) - The rate that controls the probability of bloom set/check operations being false positives. Example: 0.001 means 1 in every 1000 unique item adds/checks can be a false positive.
+
+## Use cases / Memory Savings
+
+In this example, we are simulating a very common use case of bloom filters: Ad / Event Deduplication. Applications can utilize bloom filters to track whether an advertisement / event notification has already been shown to a customer and use this to prevent showing it again to the customer.
```

Comment:
Should we get rid of mentioning the "event" ? I don't see it's being used further.

```
In this example, we are simulating a very common use case of bloom filters: advertisement view deduplication. Applications can utilize bloom filters to track whether an advertisement has already been shown to a customer and use this to prevent showing it again to the customer.
```